geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java

```diff
@@ -1411,11 +1411,15 @@ private void close(String reason, boolean cleanupEndpoint, boolean p_removeEndpo
     // This cancels the idle timer task, but it also removes the tasks reference to this connection,
     // freeing up the connection (and it's buffers for GC sooner.
     if (idleTask != null) {
-      idleTask.cancel();
+      synchronized (idleTask) {
```

Comment:
I see `idleTask` is never changed from non-null back to `null` ✓ OTOH since the variable is neither `volatile` nor guarded when we do the null-check, it is possible that if `setIdleTimeoutTask()` was called from a different thread than `close()` we could leak tasks.

I feel like maybe `idleTask` and `ackTimeoutTask` need to be volatile for this reason. Or maybe they need to be `final` and set from the constructor.

Reply:
It's likely that these variables are set in one thread and then accessed in another, so I've followed your advice and made them volatile.  I've also added checks in the two timer tasks to cancel themselves if they see that the Connection they're monitoring has closed.  That should stop any remaining leaks.